pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/partition/StreamPartitionAssignmentGenerator.java

```diff
@@ -0,0 +1,113 @@
+ * instances and all tables needing reassignment
+ */
+public class StreamPartitionAssignmentGenerator {
+
+  private ZkHelixPropertyStore<ZNRecord> _propertyStore;
+
+  public StreamPartitionAssignmentGenerator(ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    _propertyStore = propertyStore;
+  }
+
+  /**
+   * Creates a stream partition assignment from the partition assignment znode for the given realtime table
+   * @param realtimeTableName
+   * @return
+   */
+  public PartitionAssignment getStreamPartitionAssignment(String realtimeTableName) {
+    PartitionAssignment partitionAssignment = null;
+    ZNRecord streamPartitionAssignment = getStreamPartitionAssignmentFromPropertyStore(realtimeTableName);
+    if (streamPartitionAssignment != null) {
+      partitionAssignment = new PartitionAssignment(realtimeTableName, streamPartitionAssignment.getListFields());
+    }
+    return partitionAssignment;
+  }
+
+  private ZNRecord getStreamPartitionAssignmentFromPropertyStore(final String realtimeTableName) {
+    final String path = ZKMetadataProvider.constructPropertyStorePathForKafkaPartitions(realtimeTableName);
+    return _propertyStore.get(path, null, AccessOption.PERSISTENT);
+  }
+
+  /**
+   * Generates partition assignment for given table, given num partitions over given instances
+   */
+  public Map<String, PartitionAssignment> generatePartitionAssignment(TableConfig tableConfig, int numPartitions,
+      List<String> instanceNames, List<String> allTables) {
+
+    // get table configs for all tables in same tenant
+    List<TableConfig> allTableConfigs = new ArrayList<>(allTables.size());
+
+    // get current partition assignments for all tables in same tenant
+    Map<String, PartitionAssignment> tableNameToPartitionAssignment = new HashMap<>(allTables.size());
+
+    for (String tableName : allTables) {
+      allTableConfigs.add(getRealtimeTableConfig(tableName));
+      PartitionAssignment partitionAssignment = getStreamPartitionAssignment(tableName);
+      if (partitionAssignment != null) {
+        tableNameToPartitionAssignment.put(tableName, partitionAssignment);
+      }
+    }
+
+    StreamPartitionAssignmentStrategy streamPartitionAssignmentStrategy =
+        StreamPartitionAssignmentStrategyFactory.getStreamPartitionAssignmentStrategy(tableConfig);
+
+    streamPartitionAssignmentStrategy.init(allTableConfigs, instanceNames, tableNameToPartitionAssignment);
+    Map<String, PartitionAssignment> newPartitionAssignment = streamPartitionAssignmentStrategy.
+        generatePartitionAssignment(tableConfig, numPartitions);
+
+    return newPartitionAssignment;
+  }
+
+  @VisibleForTesting
+  protected TableConfig getRealtimeTableConfig(String tableNameWithType) {
+    return ZKMetadataProvider.getTableConfig(_propertyStore, tableNameWithType);
+  }
+
+  /**
+   * Given map of table name to stream partition assignment, construct and write znodes to property store one by one
+   * @param newPartitionAssignment
```

Comment:
pls add a comment that the key in the map should be raw table name, not tablename with type.

Reply:
It is actually the other way. Everything here operates assuming tableNameWithType